src/utils.py

```diff
@@ -59,6 +60,48 @@ def do_build_all_images():
             and JobParameters.py_versions == constants.ALL
         )
 
+def download_s3_file(bucket_name, filepath, local_file_name):
+    """
+
+    :param bucket_name: string
+    :param filepath: string
+    :param local_file_name: string
+    :return:
+    """
+    _s3 = boto3.Session().resource('s3')
+
+    try:
+        _s3.Bucket(bucket_name).download_file(filepath, local_file_name)
+    except ClientError as e:
+        LOGGER.error("Error: Cannot read file from s3 bucket.")
+        LOGGER.error("Exception: {}".format(e))
+        raise
+
+def download_file(remote_url: str, link_type: str):
+    """
+    Fetch remote files and save with provided local_path name
+    :param link_type: string
+    :param remote_url: string
+    :return: file_name: string
+    """
+    LOGGER.info(f"Downloading {remote_url}")
+
+    ctx = Context()
+    file_name = ctx.run(f"basename {remote_url}").stdout.strip()
+    LOGGER.info(f"basename: {file_name}")
+
+    if link_type in ["s3"] and remote_url.startswith("s3://"):
+        match = re.match(r's3://(.+?)/(.+)', remote_url)
+        bucket_name = match.group(1)
```

Comment:
Add escape characters to "/"s in the regular expression.

Reply:
Also, how is `(.+?)` different from `(.+)`?

Reply:
Forward slashes needn't be escaped. IDE also recommends it as redundant. However, for generic representation in line with other languages, I've added them back. 

(.+?) will match up to one entry between slashes, that'll form the 'bucket', the rest becomes the key, including slashes.

Reply:
Thanks. I saw recommendations to escape forward slashes, but if it is redundant, then that's fine.

Sure. `(.+?)` will match with a bucket name, but why do we need the `?`? Doesn't it signify that a particular character/group is optional? (it needs to occur between 0 and 1 times)

Reply:
In this case - the bucket name is expected to exist and will match up to 1 occurrences of the bucket name.